backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/VmHandler.java

```diff
@@ -1324,9 +1324,15 @@ public ValidationResult validateSmartCardDevice(VmBase parametersStaticData) {
     }
 
     public void updateCpuAndNumaPinning(VM vm, Guid vdsId) {
-        if (vm.getCpuPinningPolicy() == CpuPinningPolicy.RESIZE_AND_PIN_NUMA && vm.getCurrentCpuPinning() == null) {
+        if (vm.getCpuPinningPolicy() == CpuPinningPolicy.RESIZE_AND_PIN_NUMA) {
             VdsDynamic host = vdsDynamicDao.get(vdsId);
-            NumaPinningHelper.applyAutoPinningPolicy(vm, host, vdsNumaNodeDao.getAllVdsNumaNodeByVdsId(host.getId()));
+            List<VdsNumaNode> hostNumaNodes = vdsNumaNodeDao.getAllVdsNumaNodeByVdsId(host.getId());
+
+            if (vm.getCurrentCpuPinning() == null) {
+                NumaPinningHelper.applyAutoPinningPolicy(vm, host, hostNumaNodes);
+            } else {
+                vm.setCurrentCpuPinning(NumaPinningHelper.getSapHanaCpuPinning(vm, host, hostNumaNodes));
```

Comment:
do you wish to change the numa pinning as well according to the host?

Reply:
no, I think that would go against the premise that the numa pinning and cpu counts do not change during the migration, right? Just the CPU ids might be different on different hosts.

Reply:
yeah. that makes sense. but i do fear from some edge cases where the required cpus doesn't really fit to the numa we now going to use somehow. the algorithm should work for that, but worth seeing if it breaks somehow.

Reply:
> yeah. that makes sense. but i do fear from some edge cases where the required cpus doesn't really fit to the numa we now going to use somehow. the algorithm should work for that, but worth seeing if it breaks somehow.

right but how would you do that? :)

so now that we clear the 'current' pinning when suspending a vm this would become irrelevant for suspect/resume flow, right? and for migrate vm we handled it elsewhere, no?

Reply:
> > the algorithm should work for that, but worth seeing if it breaks somehow.
> 
> right but how would you do that? :)
> 

probably using unit tests 
> so now that we clear the 'current' pinning when suspending a vm this would become irrelevant for suspect/resume flow, right? and for migrate vm we handled it elsewhere, no?

i don't think we change anything when migrating (only when dealing with exclusive pinning).

Reply:
> > > the algorithm should work for that, but worth seeing if it breaks somehow.
> > 
> > 
> > right but how would you do that? :)
> 
> probably using unit tests

:)
of course but unless you use something fancy that generates inputs and validates the correctness of the output, you need to come up with inputs that reflect scenarios that may be problematic in the real-world beforehand .. if you remember, we've faced handful of issues that originated from inputs we didn't expect to get from real hosts or outputs that were reported as sub-optimal and based on them you've added unit tests..

> > so now that we clear the 'current' pinning when suspending a vm this would become irrelevant for suspect/resume flow, right? and for migrate vm we handled it elsewhere, no?
> 
> i don't think we change anything when migrating (only when dealing with exclusive pinning).

I'd put it differently - based on what @liranr23 and I talked offline earlier today, the ability to adjust the pinning when migrating the vm was only used for changing exclusive pinning and that contradicts what I thought before - that we also too care of such vms with resize-and-pin

Reply:
> > > > the algorithm should work for that, but worth seeing if it breaks somehow.
> > > 
> > > 
> > > right but how would you do that? :)
> > 
> > 
> > probably using unit tests
> 
> :) of course but unless you use something fancy that generates inputs and validates the correctness of the output, you need to come up with inputs that reflect scenarios that may be problematic in the real-world beforehand .. if you remember, we've faced handful of issues that originated from inputs we didn't expect to get from real hosts or outputs that were reported as sub-optimal and based on them you've added unit tests..
> 
> > > so now that we clear the 'current' pinning when suspending a vm this would become irrelevant for suspect/resume flow, right?
> 
> as this part was not addressed above, I'd explain what bothers me with this - if we clear the 'current' cpu pinning when stopping a suspended vm (which seems good to me) and consequently we'll call NumaPinningHelper.applyAutoPinningPolicy here, that would be problematic - we shouldn't change the cpu topology of a vm that is being resumed (it is similar to a vm that migrates)
> 
That is a good point. Should we prevent the VM starting on a destination host if the numa topology
is different instead? 

> > > and for migrate vm we handled it elsewhere, no?
> > 
> > 
> > i don't think we change anything when migrating (only when dealing with exclusive pinning).
> 
> I'd put it differently then - based on an offline conversation I had with @liranr23 earlier today, the ability to adjust the pinning when migrating the vm was only used for changing exclusive pinning and that contradicts what I thought before - that we also took care of such vms with resize-and-pin.. so would this configuration even propagate to VDSM in this case?

Reply:
> > :) of course but unless you use something fancy that generates inputs and validates the correctness of the output, you need to come up with inputs that reflect scenarios that may be problematic in the real-world beforehand .. if you remember, we've faced handful of issues that originated from inputs we didn't expect to get from real hosts or outputs that were reported as sub-optimal and based on them you've added unit tests..
> > > > so now that we clear the 'current' pinning when suspending a vm this would become irrelevant for suspect/resume flow, right?
> > 
> > 
> > as this part was not addressed above, I'd explain what bothers me with this - if we clear the 'current' cpu pinning when stopping a suspended vm (which seems good to me) and consequently we'll call NumaPinningHelper.applyAutoPinningPolicy here, that would be problematic - we shouldn't change the cpu topology of a vm that is being resumed (it is similar to a vm that migrates)
> 
> That is a good point. Should we prevent the VM starting on a destination host if the numa topology is different instead?

in theory it should be similar to live migration - and allow resuming the vm on hosts that have more cpu resources and this vm can run on. but unlike live-migration, where the vm would still run on the source in case of a failure to start the vm, here we'd probably clear the memory volumes and lost its memory state. so I think preventing that as you suggests is a reasonable solution for this